lib/model/query/datasets.js

```diff
@@ -0,0 +1,241 @@
+// Copyright 2022 ODK Central Developers
+// See the NOTICE file at the top-level directory of this distribution and at
+// https://github.com/getodk/central-backend/blob/master/NOTICE.
+// This file is part of ODK Central. It is subject to the license terms in
+// the LICENSE file found in the top-level directory of this distribution and at
+// https://www.apache.org/licenses/LICENSE-2.0. No part of ODK Central,
+// including this file, may be copied, modified, propagated, or distributed
+// except according to the terms contained in the LICENSE file.
+
+const { sql } = require('slonik');
+const { raw } = require('slonik-sql-tag-raw');
+const { unjoiner } = require('../../util/db');
+const { Dataset, Form } = require('../frames');
+const { map, reduce, compose, pickBy, startsWith, nthArg, assoc, keys, curry, nth, isEmpty, isNil, either, reduceBy } = require('ramda');
+const { construct } = require('../../util/util');
+
+// It removes prefix from all the key of an object
+const removePrefix = curry((prefix, obj) => compose(reduce((acc, key) => assoc(key.replace(prefix, ''), obj[key], acc), {}), keys)(obj));
+
+const pickFrameFields = (frame, obj) => compose(
+  removePrefix(`${frame.def.from}!`),
+  pickBy(compose(startsWith(`${frame.def.from}!`), nthArg(1)))
+)(obj);
+
+const makeHierarchy = reduce((result, item) => {
+  const dataset = new Dataset(pickFrameFields(Dataset, item)).forApi();
+  const property = new Dataset.Property(pickFrameFields(Dataset.Property, item)).forApi();
+  const propertyField = new Dataset.PropertyField(pickFrameFields(Dataset.PropertyField, item)).forApi();
+
+  return {
+    ...result,
+    [dataset.id]: {
+      ...dataset,
+      properties: !property.id ? { ...(result[dataset.id]?.properties || {}) } : {
+        ...(result[dataset.id]?.properties || {}),
+        [property.id]: {
+          ...property,
+          fields: [
+            ...(result[dataset.id]?.properties[property.id]?.fields || []),
+            propertyField
+          ]
```

Comment:
I don't see this used anywhere? But maybe I missed something?

Reply:
removed in #697